Automattic/jetpack
# Add Jetpack WAF standalone mode

Introduces the standalone mode to the Jetpack WAF. This means that the firewall is run in front of WordPress by making use of the [PHP auto_prepend_file config directive](https://www.php.net/manual/en/ini.core.php#ini.auto-prepend-file). Guiding the user how to configure this is not part of this PR.

#### Changes proposed in this Pull Request:

* Add generator to generate WAF standalone mode bootstrap.
* Add CLI command to set the mode (this triggers the regeneration of the bootstrap).

#### Jetpack product discussion



1201069996155224-as-1201782715336899

#### Does this pull request change what data or activity we track or use?



No.

#### Testing instructions:






It is a bit complicated because it is not clearly visible from the outside whether the firewall blocks a request in plugin mode or in standalone mode.

* Check out this branch
* Add something that you will recognize to the `plugin_loaded` action before(!) including the run.php, for example: `phpinfo(); die;`
* Verify that you see the phpinfo (or what else you added) in your browser when accessing the page with the `?jetpack-waf-always-block` query parameter
* Set a mode to trigger the generation of the bootstrap file: `wp jetpack-waf mode normal`
* Do all of the indented ones below inside the docker container: `docker exec -it jetpack_dev_wordpress_1 bash`
  * configure the auto-prepend: `echo "auto_prepend_file=/var/www/html/wp-content/jetpack-waf/bootstrap.php" >> /etc/php/8.0/apache2/conf.d/30-jpwaf.ini`
  * restart apache to update config: `apachectl graceful` 
* Verify that your request is now blocked and you see the rule 1337 block.
  * remove config again: `rm /etc/php/8.0/apache2/conf.d/30-jpwaf.ini`
  * restart apache to update config: `apachectl graceful` 
* Verify you see the phpinfo again

#### Other notes

- I did not find a decent way to add tests for the CLI command, which is a bummer. I think we could add [AspectMock](https://github.com/Codeception/AspectMock) to mock away the static calls? Glad to hear other suggestions.
- Also, I did not add generating the bootstrap upon plugin activation, because all the code is inside this package right now. I think the proper way to do this would be to actually call for the generation from the jetpack plugin itself? Happy to add that as a separate PR if I am on the right path with my thoughts here, but I just was not sure.
- I took the `get_option` mock for the tests from the jetpack-autoloader package, this seemed like the most reasonable solution.
- I'll re-require the things in the jetpack plugin once the reviews are through to avoid doing this multiple times.
